apps/sync/js/adapters/bookmarks.js

```diff
@@ -0,0 +1,334 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+  Bookmarks Data Adapter
+
+    The purpose of the Bookmarks adapter is to retrieve bookmark records from
+  Firefox Sync and merge them in a local bookmarks DataStore
+  ('firefox-sync-bookmarks') which will be hosting Firefox Sync records only.
+  This adapter implements the API defined by SyncEngine. For now the
+  implementation allows only to retrieve bookmarks from Sync but it is not ready
+  for pushing any changes. The high level algorithm implemented here is the
+  following:
+  - Traverse all the records coming from Firefox Sync and stored locally
+    in Kinto.
+  - Convert any new bookmark from the Firefox Sync format to the
+    'firefox-sync-bookmarks' DataStore one.
+  - For existing bookmarks merge the remote with the local bookmark after
+    converting them to the proper 'firefox-sync-bookmarks' DS format.
+  - Remove any existing bookmark that has been remotely deleted (these records
+    come with a 'deleted: true' property).
+**/
+
+'use strict';
+
+/* global
+  asyncStorage,
+  DataAdapters,
+  LazyLoader
+*/
+
+const BOOKMARKS_COLLECTION_MTIME = 'collections::bookmarks::mtime';
+const BOOKMARKS_SYNCTOID_PREFIX = 'synctoid::bookmarks::';
+
+var BookmarksHelper = (() => {
+  var store;
+  function _ensureStore() {
+    if (store) {
+      return Promise.resolve(store);
+    }
+    return navigator.getDataStores('firefox-sync-bookmarks').then(stores => {
+      store = stores[0];
+      return store;
+    });
+  }
+
+  function setSyncedCollectionMtime(mtime) {
+    return new Promise(resolve => {
+      asyncStorage.setItem(BOOKMARKS_COLLECTION_MTIME, mtime, resolve);
```

Comment:
This timestamp is for Bookmarks Collection rather than DataStore, and it's necessary to be recorded what records had been synchronized.